File: packages/core-ts/src/oauth/user-store.ts
/**
 * Per-user encrypted credential store for MCP HTTP modes.
 * Uses better-sqlite3 for persistence and Node.js crypto for AES-256-GCM at-rest encryption.
 */

import { createCipheriv, createDecipheriv, randomBytes } from 'node:crypto'
import { mkdirSync } from 'node:fs'
import { dirname } from 'node:path'
import Database from 'better-sqlite3'

export interface IUserCredentialStore {
  saveCredentials(userId: string, config: Record<string, string>): void
  getCredentials(userId: string): Record<string, string> | null
  deleteCredentials(userId: string): void
}

export class SqliteUserStore implements IUserCredentialStore {
  private db: Database.Database
  private masterKey: Buffer

  /**
   * @param dbPath Path to the SQLite DB file.
   * @param masterKey 32-byte AES key for at-rest encryption.
   */
  constructor(dbPath: string, masterKey: Buffer) {
    if (masterKey.length !== 32) throw new Error('masterKey must be 32 bytes')
    this.masterKey = masterKey

    // Ensure directory exists
    mkdirSync(dirname(dbPath), { recursive: true })

    this.db = new Database(dbPath)
    this.db.pragma('journal_mode = WAL')
    this.db.exec(`
      CREATE TABLE IF NOT EXISTS users (
        user_id TEXT PRIMARY KEY,
        encrypted_config BLOB NOT NULL,
        created_at INTEGER NOT NULL,
        updated_at INTEGER NOT NULL
      )
    `)
  }

  private encrypt(plaintext: string): Buffer {
    const iv = randomBytes(12)
    const cipher = createCipheriv('aes-256-gcm', this.masterKey, iv, { authTagLength: 16 })
    const encrypted = Buffer.concat([cipher.update(plaintext, 'utf-8'), cipher.final()])
    const tag = cipher.getAuthTag()
    // Layout: [12 iv] [16 tag] [... ciphertext]
    return Buffer.concat([iv, tag, encrypted])
  }

  private decrypt(payload: Buffer): string {
    const iv = payload.subarray(0, 12)
    const tag = payload.subarray(12, 28)
    const ciphertext = payload.subarray(28)
    const decipher = createDecipheriv('aes-256-gcm', this.masterKey, iv, { authTagLength: 16 })
    decipher.setAuthTag(tag)
    return Buffer.concat([decipher.update(ciphertext), decipher.final()]).toString('utf-8')
  }

  saveCredentials(userId: string, config: Record<string, string>): void {
    const now = Math.floor(Date.now() / 1000)
    const encrypted = this.encrypt(JSON.stringify(config))

    this.db
      .prepare(`
        INSERT INTO users (user_id, encrypted_config, created_at, updated_at)
        VALUES (?, ?, ?, ?)
        ON CONFLICT(user_id) DO UPDATE SET
          encrypted_config = excluded.encrypted_config,
          updated_at = excluded.updated_at
      `)
      .run(userId, encrypted, now, now)
  }

  getCredentials(userId: string): Record<string, string> | null {
    const row = this.db.prepare('SELECT encrypted_config FROM users WHERE user_id = ?').get(userId) as
      | { encrypted_config: Buffer }
      | undefined

    if (!row) return null

    try {
      return JSON.parse(this.decrypt(Buffer.from(row.encrypted_config)))
    } catch {
      return null
    }
  }

  deleteCredentials(userId: string): void {
    this.db.prepare('DELETE FROM users WHERE user_id = ?').run(userId)
  }

  close(): void {
    this.db.close()
  }
}

File: packages/core-py/src/mcp_relay_core/oauth/user_store.py
"""Per-user encrypted credential store for MCP HTTP modes."""

import json
import os
import sqlite3
from abc import ABC, abstractmethod
from pathlib import Path

from cryptography.hazmat.primitives.ciphers.aead import AESGCM


class IUserCredentialStore(ABC):
    @abstractmethod
    def save_credentials(self, user_id: str, config: dict) -> None:
        pass

    @abstractmethod
    def get_credentials(self, user_id: str) -> dict | None:
        pass

    @abstractmethod
    def delete_credentials(self, user_id: str) -> None:
        pass


class SqliteUserStore(IUserCredentialStore):
    def __init__(self, db_path: Path, master_key: bytes):
        """
        Args:
            db_path: Path to the SQLite DB file.
            master_key: 32-byte AES key for at-rest encryption.
        """
        self.db_path = db_path
        self._master_key = master_key

        # Ensure directory exists with strict permissions
        self.db_path.parent.mkdir(parents=True, exist_ok=True)
        if self.db_path.parent.exists() and os.name != "nt":
            os.chmod(self.db_path.parent, 0o700)  # nosemgrep

        self._init_db()

    def _init_db(self) -> None:
        with sqlite3.connect(self.db_path) as conn:
            conn.execute(
                """
                CREATE TABLE IF NOT EXISTS users (
                    user_id TEXT PRIMARY KEY,
                    encrypted_config BLOB NOT NULL,
                    created_at INTEGER NOT NULL,
                    updated_at INTEGER NOT NULL
                )
                """
            )

    def _encrypt(self, plaintext: str) -> bytes:
        iv = os.urandom(12)
        aesgcm = AESGCM(self._master_key)
        # return IV + combined(ciphertext+tag)
        combined = aesgcm.encrypt(iv, plaintext.encode("utf-8"), None)
        return iv + combined

    def _decrypt(self, payload: bytes) -> str:
        iv = payload[:12]
        combined = payload[12:]
        aesgcm = AESGCM(self._master_key)
        plaintext_bytes = aesgcm.decrypt(iv, combined, None)
        return plaintext_bytes.decode("utf-8")

    def save_credentials(self, user_id: str, config: dict) -> None:
        import time

        now = int(time.time())
        json_str = json.dumps(config)
        encrypted = self._encrypt(json_str)

        with sqlite3.connect(self.db_path) as conn:
            conn.execute(
                """
                INSERT INTO users (user_id, encrypted_config, created_at, updated_at)
                VALUES (?, ?, ?, ?)
                ON CONFLICT(user_id) DO UPDATE SET
                    encrypted_config=excluded.encrypted_config,
                    updated_at=excluded.updated_at
                """,
                (user_id, encrypted, now, now),
            )

    def get_credentials(self, user_id: str) -> dict | None:
        with sqlite3.connect(self.db_path) as conn:
            cursor = conn.execute(
                "SELECT encrypted_config FROM users WHERE user_id = ?", (user_id,)
            )
            row = cursor.fetchone()

        if not row:
            return None

        try:
            json_str = self._decrypt(row[0])
            return json.loads(json_str)
        except Exception:
            return None

    def delete_credentials(self, user_id: str) -> None:
        with sqlite3.connect(self.db_path) as conn:
            conn.execute("DELETE FROM users WHERE user_id = ?", (user_id,))
